nmigen_boards/resources/interface.py

```diff
@@ -81,3 +84,35 @@ def SPIResource(*args, cs, clk, copi, cipo, int=None, reset=None,
     if attrs is not None:
         io.append(attrs)
     return Resource.family(*args, default_name="spi", ios=io)
+
+
+def DirectUSBResource(*args, dp, dn, pullup=None, vbus_valid=None,
+    conn=None, attrs=None):
+
+    io = []
+    io.append(Subsignal("d_p", Pins(dp, dir="io", conn=conn, assert_width=1)))
+    io.append(Subsignal("d_n", Pins(dn, dir="io", conn=conn, assert_width=1)))
+    if pullup:
+        io.append(Subsignal("pullup", Pins(pullup, dir="o", conn=conn, assert_width=1)))
+    if vbus_valid:
+        io.append(Subsignal("vbus_valid", Pins(pullup, dir="i", conn=conn, assert_width=1)))
```

Comment:
Typo
```suggestion
        io.append(Subsignal("vbus_valid", Pins(vbus_valid, dir="i", conn=conn, assert_width=1)))
```

I'd also probably call it something like `vbus_detect` but this isn't important.

Reply:
_VbusValid_ is what UTMI/ULPI call it, so it makes sense not to come up with our own name. (It's ideally to be the output of a level comparator, even if it's often not in practice.)

Reply:
Sounds good then.